src/jobserver-posix.cc

```diff
@@ -0,0 +1,160 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "jobserver.h"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <cstring>
+
+#include "util.h"
+
+// Declare complete type for static constants in class.
+constexpr char const Jobserver::kAuthKey[];
+constexpr char const Jobserver::kFdsKey[];
+constexpr char const Jobserver::kFifoKey[];
+
+PosixJobserverClient::PosixJobserverClient() {
+  assert(!Enabled());
+
+  // Set name, type of pipe, and if non-parallel from MAKEFLAGS.
+  Parse();
+
+  const char* jobserver = jobserver_name_.c_str();
+
+  // Warn if jobserver type is unknown (neither fifo nor pipe).
+  if (!jobserver_fifo_ && sscanf(jobserver, "%d,%d", &rfd_, &wfd_) != 2)
+    if (!jobserver_name_.empty())
+      Warning("invalid jobserver auth: '%s'", jobserver);
+
+  // Open FDs to the pipe if needed, read must be non-blocking.
+  // If passed FDs are blocking on read, force non-parallel build.
+  if (jobserver_fifo_) {
+    rfd_ = open(jobserver + strlen(kFifoKey), O_RDONLY | O_NONBLOCK);
+    wfd_ = open(jobserver + strlen(kFifoKey), O_WRONLY);
+  } else if (Enabled() && (fcntl(rfd_, F_GETFL) & O_NONBLOCK) == 0) {
+    jobserver_closed_ = true;
+  }
+
+  // Exit on failure to open FDs, build non-parallel for invalid passed FDs.
+  if (Enabled())
+    Info("using jobserver: %s", jobserver);
+  else if (jobserver_fifo_ && (rfd_ == -1 || wfd_ == -1))
+    Fatal("failed to open jobserver: %s: %s", jobserver, strerror(errno));
+  else if (!jobserver_name_.empty())
+    jobserver_closed_ = true;
+
+  // Signal that we have initialized but do not have a token yet.
+  if (Enabled())
+    token_count_ = -1;
+}
+
+void PosixJobserverClient::Parse() {
+  // Return early if no makeflags are passed in the environment.
+  const char* makeflags = std::getenv("MAKEFLAGS");
+  if (makeflags == nullptr || strlen(makeflags) == 0)
+    return;
+
+  std::string::size_type flag_char = 0;
+  std::string flag;
+  std::vector<std::string> flags;
+
+  // Tokenize string to characters in flag, then words in flags.
+  while (flag_char < strlen(makeflags)) {
+    while (flag_char < strlen(makeflags) &&
+           !isblank(static_cast<unsigned char>(makeflags[flag_char]))) {
+      flag.push_back(static_cast<unsigned char>(makeflags[flag_char]));
+      flag_char++;
+    }
+
+    if (!flag.empty())
+      flags.push_back(flag);
+
+    flag.clear();
+    flag_char++;
+  }
+
+  // --jobserver-auth=<val>
+  for (size_t n = 0; n < flags.size(); n++)
```

Comment:
these loops look like they're missing break;

Reply:
lack of `break` is intentional, so only the last matching value is the one that takes effect

Reply:
Sounds like it would be better to reverse the loop then.

Reply:
sure it can be reversed, but since this will only be run once per execution of ninja it's not performance critical, so I lean towards the simplest way to write it with less lines and no brackets, etc...